File: bin/semantic-ex/gen-schemas.py
#!/usr/bin/env python3
"""
Observe fake data and generate JSON Schemas (semantic-ex Phase 2 + 3).
maxLength = observed max * ~1.5 rounded up; minLength = observed min.
"""
import json
import math
from pathlib import Path

FAKE = Path("/Users/akihito/git/BEAR.Cms/var/fake")
OUT  = Path("/Users/akihito/git/BEAR.Cms/var/json_schema")
OUT.mkdir(parents=True, exist_ok=True)

BASE_ID = "https://example.com/bear-cms"

def ceil_to_nice(n):
    # round max up to a nice boundary (like 50, 100, 200, 500, 1000, ...)
    if n <= 50: return 50
    if n <= 100: return 100
    if n <= 200: return 200
    if n <= 500: return 500
    if n <= 1000: return 1000
    if n <= 2000: return 2000
    if n <= 5000: return 5000
    if n <= 10000: return 10000
    return int(math.ceil(n * 1.5 / 1000)) * 1000

def observe_str(values):
    nonnull = [v for v in values if v is not None]
    nulls = len(values) - len(nonnull)
    lengths = [len(str(v)) for v in nonnull]
    return {
        "nulls": nulls,
        "min_len": min(lengths) if lengths else 0,
        "max_len": max(lengths) if lengths else 0,
        "typical_len": sorted(lengths)[len(lengths)//2] if lengths else 0,
    }

def observe_int(values):
    nonnull = [v for v in values if v is not None]
    nulls = len(values) - len(nonnull)
    return {
        "nulls": nulls,
        "min": min(nonnull) if nonnull else 0,
        "max": max(nonnull) if nonnull else 0,
    }

def field_list(records, key):
    return [r.get(key) for r in records]

def str_field_schema(records, key, *, description, format_=None, pattern=None, nullable=False):
    obs = observe_str(field_list(records, key))
    if obs["max_len"] == 0 and obs["min_len"] == 0 and obs["nulls"] > 0:
        min_len = 0
        max_len = 100
    else:
        min_len = obs["min_len"]
        max_len = ceil_to_nice(obs["max_len"])
    schema = {
        "type": ["string", "null"] if nullable else "string",
        "description": description,
        "minLength": min_len,
        "maxLength": max_len,
    }
    if format_:
        schema["format"] = format_
    if pattern:
        schema["pattern"] = pattern
    return schema

def int_field_schema(records, key, *, description, nullable=False, minimum=None):
    obs = observe_int(field_list(records, key))
    schema = {
        "type": ["integer", "null"] if nullable else "integer",
        "description": description,
    }
    if minimum is not None:
        schema["minimum"] = minimum
    else:
        schema["minimum"] = obs["min"]
    return schema

# -- load fake data -------------------------------------------------------
articles   = json.loads((FAKE / "article.json").read_text())
categories = json.loads((FAKE / "category.json").read_text())
tags       = json.loads((FAKE / "tag.json").read_text())
authors    = json.loads((FAKE / "author.json").read_text())
media      = json.loads((FAKE / "media.json").read_text())

link_def = {
    "type": "object",
    "required": ["href"],
    "properties": {
        "href": {"type": "string", "format": "uri-reference"}
    }
}

# -- Article schema -------------------------------------------------------
article_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/article",
    "title": "Article",
    "description": "ALPS Article state. Single article with embedded Author, Category and Tags.",
    "type": "object",
    "required": ["id", "slug", "title", "body", "status", "authorId", "categoryId"],
    "properties": {
        "id": int_field_schema(articles, "id", description="Primary key", minimum=1),
        "slug": str_field_schema(articles, "slug", description="URL-safe unique slug", pattern="^[a-z0-9][a-z0-9-]*$"),
        "title": str_field_schema(articles, "title", description="Article headline"),
        "body": str_field_schema(articles, "body", description="Full article body"),
        "excerpt": str_field_schema(articles, "excerpt", description="Short summary used in list views"),
        "status": {
            "type": "string",
            "description": "Lifecycle status",
            "enum": ["draft", "published"]
        },
        "publishedAt": {
            "type": ["string", "null"],
            "description": "Publication timestamp; null for drafts",
            "format": "date-time"
        },
        "authorId": int_field_schema(articles, "authorId", description="Foreign key to Author", minimum=1),
        "categoryId": int_field_schema(articles, "categoryId", description="Foreign key to Category", minimum=1),
        "_links": {
            "type": "object",
            "description": "HAL links. goArticleList always present; doUpdateArticle/doDeleteArticle for write operations.",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goArticleList": {"$ref": "#/$defs/link"},
                "goAuthor": {"$ref": "#/$defs/link"},
                "goCategory": {"$ref": "#/$defs/link"},
                "doUpdateArticle": {"$ref": "#/$defs/link"},
                "doDeleteArticle": {"$ref": "#/$defs/link"}
            }
        },
        "_embedded": {
            "type": "object",
            "description": "Embedded Author, Category and Tag resources",
            "properties": {
                "author": {"$ref": f"{BASE_ID}/author"},
                "category": {"$ref": f"{BASE_ID}/category"},
                "tags": {
                    "type": "array",
                    "items": {"$ref": f"{BASE_ID}/tag"}
                }
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- ArticleList schema ---------------------------------------------------
article_list_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/articleList",
    "title": "ArticleList",
    "description": "Paginated list of articles",
    "type": "object",
    "required": ["items", "page", "perPage", "totalCount"],
    "properties": {
        "items": {
            "type": "array",
            "items": {"$ref": f"{BASE_ID}/article"}
        },
        "page": {"type": "integer", "minimum": 1, "description": "Current page (1-indexed)"},
        "perPage": {"type": "integer", "minimum": 1, "maximum": 100, "description": "Items per page"},
        "totalCount": {"type": "integer", "minimum": 0, "description": "Total matching articles"},
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "next": {"$ref": "#/$defs/link"},
                "prev": {"$ref": "#/$defs/link"},
                "first": {"$ref": "#/$defs/link"},
                "last": {"$ref": "#/$defs/link"},
                "goArticle": {"$ref": "#/$defs/link"},
                "doCreateArticle": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- Category schema ------------------------------------------------------
category_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/category",
    "title": "Category",
    "description": "ALPS Category state",
    "type": "object",
    "required": ["id", "slug", "name"],
    "properties": {
        "id": int_field_schema(categories, "id", description="Primary key", minimum=1),
        "slug": str_field_schema(categories, "slug", description="URL-safe unique slug", pattern="^[a-z0-9][a-z0-9-]*$"),
        "name": str_field_schema(categories, "name", description="Display name"),
        "description": str_field_schema(categories, "description", description="Optional description", nullable=True),
        "parentId": int_field_schema(categories, "parentId", description="Parent category id, null for top-level", nullable=True, minimum=1),
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goCategoryList": {"$ref": "#/$defs/link"},
                "goArticleList": {"$ref": "#/$defs/link"},
                "doUpdateCategory": {"$ref": "#/$defs/link"},
                "doDeleteCategory": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- CategoryList schema --------------------------------------------------
category_list_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/categoryList",
    "title": "CategoryList",
    "description": "List of categories",
    "type": "object",
    "required": ["items"],
    "properties": {
        "items": {"type": "array", "items": {"$ref": f"{BASE_ID}/category"}},
        "totalCount": {"type": "integer", "minimum": 0},
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goCategory": {"$ref": "#/$defs/link"},
                "doCreateCategory": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- Tag schema -----------------------------------------------------------
tag_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/tag",
    "title": "Tag",
    "description": "ALPS Tag state",
    "type": "object",
    "required": ["id", "slug", "name"],
    "properties": {
        "id": int_field_schema(tags, "id", description="Primary key", minimum=1),
        "slug": str_field_schema(tags, "slug", description="URL-safe unique slug", pattern="^[a-z0-9][a-z0-9-]*$"),
        "name": str_field_schema(tags, "name", description="Display name"),
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goTagList": {"$ref": "#/$defs/link"},
                "goArticleList": {"$ref": "#/$defs/link"},
                "doDeleteTag": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- TagList schema -------------------------------------------------------
tag_list_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/tagList",
    "title": "TagList",
    "description": "List of tags",
    "type": "object",
    "required": ["items"],
    "properties": {
        "items": {"type": "array", "items": {"$ref": f"{BASE_ID}/tag"}},
        "totalCount": {"type": "integer", "minimum": 0},
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goTag": {"$ref": "#/$defs/link"},
                "doCreateTag": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- Author schema --------------------------------------------------------
author_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/author",
    "title": "Author",
    "description": "ALPS Author state",
    "type": "object",
    "required": ["id", "name", "email"],
    "properties": {
        "id": int_field_schema(authors, "id", description="Primary key", minimum=1),
        "name": str_field_schema(authors, "name", description="Full name"),
        "email": str_field_schema(authors, "email", description="Email address", format_="email"),
        "bio": str_field_schema(authors, "bio", description="Optional biography; empty allowed"),
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "goArticleList": {"$ref": "#/$defs/link"},
                "doUpdateAuthor": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- Media schema ---------------------------------------------------------
media_schema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": f"{BASE_ID}/media",
    "title": "Media",
    "description": "ALPS Media state",
    "type": "object",
    "required": ["id", "filename", "mimeType", "url"],
    "properties": {
        "id": int_field_schema(media, "id", description="Primary key", minimum=1),
        "filename": str_field_schema(media, "filename", description="Original filename"),
        "mimeType": str_field_schema(media, "mimeType", description="MIME type"),
        "url": str_field_schema(media, "url", description="Public URL path", format_="uri-reference"),
        "alt": str_field_schema(media, "alt", description="Accessibility alt text", nullable=True),
        "width": int_field_schema(media, "width", description="Pixel width; 0 for SVG", minimum=0),
        "height": int_field_schema(media, "height", description="Pixel height; 0 for SVG", minimum=0),
        "_links": {
            "type": "object",
            "properties": {
                "self": {"$ref": "#/$defs/link"},
                "doDeleteMedia": {"$ref": "#/$defs/link"}
            }
        }
    },
    "$defs": {"link": link_def}
}

# -- Write observations notes ---------------------------------------------
notes = []
def observe_and_log(label, records, fields):
    lines = [f"\n## {label}"]
    for f in fields:
        vals = field_list(records, f)
        if all(isinstance(v, (int, type(None))) for v in vals):
            obs = observe_int(vals)
            lines.append(f"- `{f}` (int): min={obs['min']} max={obs['max']} nulls={obs['nulls']}/{len(vals)}")
        else:
            obs = observe_str(vals)
            lines.append(f"- `{f}` (str): min_len={obs['min_len']} max_len={obs['max_len']} typical={obs['typical_len']} nulls={obs['nulls']}/{len(vals)}")
    notes.extend(lines)

notes.append("# Fake data observations (semantic-ex Phase 2)")
notes.append(f"Generated from {FAKE}//*.json — 50 records per atomic entity.")
observe_and_log("Article", articles, ["id","slug","title","body","excerpt","status","publishedAt","authorId","categoryId"])
observe_and_log("Category", categories, ["id","slug","name","description","parentId"])
observe_and_log("Tag", tags, ["id","slug","name"])
observe_and_log("Author", authors, ["id","name","email","bio"])
observe_and_log("Media", media, ["id","filename","mimeType","url","alt","width","height"])

(FAKE / "observations.md").write_text("\n".join(notes) + "\n")

# -- Write schemas --------------------------------------------------------
def write(name, schema):
    (OUT / f"{name}.json").write_text(json.dumps(schema, indent=2, ensure_ascii=False) + "\n")

write("article", article_schema)
write("articleList", article_list_schema)
write("category", category_schema)
write("categoryList", category_list_schema)
write("tag", tag_schema)
write("tagList", tag_list_schema)
write("author", author_schema)
write("media", media_schema)

print("Wrote schemas to", OUT)
for p in sorted(OUT.iterdir()):
    print(" ", p.name)

File: bin/semantic-ex/gen-schemas.php
<?php

declare(strict_types=1);

/**
 * Semantic-ex Phase 2 + 3 — observe fake data and emit JSON Schemas
 * with constraints derived from the data, not chosen up-front.
 *
 * maxLength = ceil(observed_max * 1.5) snapped to a 'nice' boundary;
 * minLength = observed minimum.
 *
 * PHP port of bin/semantic-ex/gen-schemas.py.
 */

$fake = dirname(__DIR__, 2) . '/var/fake';
$out = dirname(__DIR__, 2) . '/var/json_schema';
@mkdir($out, 0755, true);

$baseId = 'https://example.com/bear-cms';

function ceilToNice(int $n): int
{
    foreach ([50, 100, 200, 500, 1000, 2000, 5000, 10000] as $boundary) {
        if ($n <= $boundary) {
            return $boundary;
        }
    }

    return (int) (ceil($n * 1.5 / 1000) * 1000);
}

/** @return array{nulls:int, min_len:int, max_len:int, typical_len:int} */
function observeStr(array $values): array
{
    $nonnull = array_filter($values, static fn ($v) => $v !== null);
    $lengths = array_map(static fn ($v) => mb_strlen((string) $v), $nonnull);
    sort($lengths);

    return [
        'nulls' => count($values) - count($nonnull),
        'min_len' => $lengths === [] ? 0 : (int) min($lengths),
        'max_len' => $lengths === [] ? 0 : (int) max($lengths),
        'typical_len' => $lengths === [] ? 0 : $lengths[intdiv(count($lengths), 2)],
    ];
}

/** @return array{nulls:int, min:int, max:int} */
function observeInt(array $values): array
{
    $nonnull = array_filter($values, static fn ($v) => $v !== null);

    return [
        'nulls' => count($values) - count($nonnull),
        'min' => $nonnull === [] ? 0 : (int) min($nonnull),
        'max' => $nonnull === [] ? 0 : (int) max($nonnull),
    ];
}

function fields(array $records, string $key): array
{
    return array_map(static fn ($r) => $r[$key] ?? null, $records);
}

/** @return array<string, mixed> */
function strFieldSchema(array $records, string $key, string $description, ?string $format = null, ?string $pattern = null, bool $nullable = false): array
{
    $obs = observeStr(fields($records, $key));
    $minLen = ($obs['max_len'] === 0 && $obs['nulls'] > 0) ? 0 : $obs['min_len'];
    $maxLen = ($obs['max_len'] === 0 && $obs['nulls'] > 0) ? 100 : ceilToNice($obs['max_len']);
    $schema = [
        'type' => $nullable ? ['string', 'null'] : 'string',
        'description' => $description,
        'minLength' => $minLen,
        'maxLength' => $maxLen,
    ];
    if ($format !== null) {
        $schema['format'] = $format;
    }

    if ($pattern !== null) {
        $schema['pattern'] = $pattern;
    }

    return $schema;
}

/** @return array<string, mixed> */
function intFieldSchema(array $records, string $key, string $description, bool $nullable = false, ?int $minimum = null): array
{
    $obs = observeInt(fields($records, $key));
    $schema = [
        'type' => $nullable ? ['integer', 'null'] : 'integer',
        'description' => $description,
        'minimum' => $minimum ?? $obs['min'],
    ];

    return $schema;
}

$articles = json_decode(file_get_contents($fake . '/article.json'), true);
$articles = array_values(array_filter(
    $articles,
    static fn (array $article): bool => ($article['slug'] ?? null) !== 'xss-regression',
));
$categories = json_decode(file_get_contents($fake . '/category.json'), true);
$tags = json_decode(file_get_contents($fake . '/tag.json'), true);
$authors = json_decode(file_get_contents($fake . '/author.json'), true);
$media = json_decode(file_get_contents($fake . '/media.json'), true);

$articleSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/article',
    'title' => 'Article',
    'description' => 'ALPS Article state. Single article with embedded Author, Category and Tags.',
    'type' => 'object',
    'required' => ['id', 'slug', 'title', 'body', 'status', 'authorId', 'categoryId'],
    'properties' => [
        'id' => intFieldSchema($articles, 'id', 'Primary key', false, 1),
        'slug' => strFieldSchema($articles, 'slug', 'URL-safe unique slug', null, '^[a-z0-9][a-z0-9-]*$'),
        'title' => strFieldSchema($articles, 'title', 'Article headline'),
        'body' => strFieldSchema($articles, 'body', 'Full article body'),
        'excerpt' => strFieldSchema($articles, 'excerpt', 'Short summary used in list views', null, null, true),
        'status' => ['type' => 'string', 'description' => 'Lifecycle status', 'enum' => ['draft', 'published']],
        'publishedAt' => ['type' => ['string', 'null'], 'description' => 'Publication timestamp; null for drafts', 'format' => 'date-time'],
        'authorId' => intFieldSchema($articles, 'authorId', 'Foreign key to Author', false, 1),
        'categoryId' => intFieldSchema($articles, 'categoryId', 'Foreign key to Category', false, 1),
    ],
];

$articleListSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/articleList',
    'title' => 'ArticleList',
    'description' => 'Paginated list of article summaries',
    'type' => 'object',
    'required' => ['items', 'page', 'perPage', 'count', 'totalCount'],
    'properties' => [
        'items' => [
            'type' => 'array',
            'items' => [
                'type' => 'object',
                'required' => ['id', 'slug', 'title', 'status'],
                'properties' => [
                    'id' => ['type' => 'integer', 'minimum' => 1],
                    'slug' => ['type' => 'string'],
                    'title' => ['type' => 'string'],
                    'excerpt' => ['type' => ['string', 'null']],
                    'status' => ['type' => 'string', 'enum' => ['draft', 'published']],
                    'publishedAt' => ['type' => ['string', 'null'], 'format' => 'date-time'],
                    'authorId' => ['type' => 'integer', 'minimum' => 1],
                    'categoryId' => ['type' => 'integer', 'minimum' => 1],
                ],
            ],
        ],
        'page' => ['type' => 'integer', 'minimum' => 1],
        'perPage' => ['type' => 'integer', 'minimum' => 1, 'maximum' => 100],
        'count' => ['type' => 'integer', 'minimum' => 0],
        'totalCount' => ['type' => 'integer', 'minimum' => 0],
    ],
];

$categorySchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/category',
    'title' => 'Category',
    'type' => 'object',
    'required' => ['id', 'slug', 'name'],
    'properties' => [
        'id' => intFieldSchema($categories, 'id', 'Primary key', false, 1),
        'slug' => strFieldSchema($categories, 'slug', 'URL-safe unique slug', null, '^[a-z0-9][a-z0-9-]*$'),
        'name' => strFieldSchema($categories, 'name', 'Display name'),
        'description' => strFieldSchema($categories, 'description', 'Optional description', null, null, true),
        'parentId' => intFieldSchema($categories, 'parentId', 'Parent category id; null for top-level', true, 1),
    ],
];

$categoryListSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/categoryList',
    'title' => 'CategoryList',
    'type' => 'object',
    'required' => ['items'],
    'properties' => [
        'items' => ['type' => 'array', 'items' => ['$ref' => 'category.json']],
        'totalCount' => ['type' => 'integer', 'minimum' => 0],
    ],
];

$tagSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/tag',
    'title' => 'Tag',
    'type' => 'object',
    'required' => ['id', 'slug', 'name'],
    'properties' => [
        'id' => intFieldSchema($tags, 'id', 'Primary key', false, 1),
        'slug' => strFieldSchema($tags, 'slug', 'URL-safe unique slug', null, '^[a-z0-9][a-z0-9-]*$'),
        'name' => strFieldSchema($tags, 'name', 'Display name'),
    ],
];

$tagListSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/tagList',
    'title' => 'TagList',
    'type' => 'object',
    'required' => ['items'],
    'properties' => [
        'items' => ['type' => 'array', 'items' => ['$ref' => 'tag.json']],
        'totalCount' => ['type' => 'integer', 'minimum' => 0],
    ],
];

$authorSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/author',
    'title' => 'Author',
    'type' => 'object',
    'required' => ['id', 'name', 'email'],
    'properties' => [
        'id' => intFieldSchema($authors, 'id', 'Primary key', false, 1),
        'name' => strFieldSchema($authors, 'name', 'Full name'),
        'email' => strFieldSchema($authors, 'email', 'Email address', 'email'),
        'bio' => strFieldSchema($authors, 'bio', 'Optional biography', null, null, true),
    ],
];

$authorListSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/authorList',
    'title' => 'AuthorList',
    'type' => 'object',
    'required' => ['items'],
    'properties' => [
        'items' => ['type' => 'array', 'items' => ['$ref' => 'author.json']],
        'totalCount' => ['type' => 'integer', 'minimum' => 0],
    ],
];

$mediaSchema = [
    '$schema' => 'https://json-schema.org/draft/2020-12/schema',
    '$id' => $baseId . '/media',
    'title' => 'Media',
    'type' => 'object',
    'required' => ['id', 'filename', 'mimeType', 'url'],
    'properties' => [
        'id' => intFieldSchema($media, 'id', 'Primary key', false, 1),
        'filename' => strFieldSchema($media, 'filename', 'Original filename'),
        'mimeType' => strFieldSchema($media, 'mimeType', 'MIME type'),
        'url' => strFieldSchema($media, 'url', 'Public URL path', 'uri-reference'),
        'alt' => strFieldSchema($media, 'alt', 'Accessibility alt text', null, null, true),
        'width' => intFieldSchema($media, 'width', 'Pixel width; 0 for SVG', false, 0),
        'height' => intFieldSchema($media, 'height', 'Pixel height; 0 for SVG', false, 0),
    ],
];

$write = static function (string $name, array $schema) use ($out): void {
    file_put_contents($out . '/' . $name, json_encode($schema, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n");
};

$write('article.json', $articleSchema);
$write('articleList.json', $articleListSchema);
$write('category.json', $categorySchema);
$write('categoryList.json', $categoryListSchema);
$write('tag.json', $tagSchema);
$write('tagList.json', $tagListSchema);
$write('author.json', $authorSchema);
$write('authorList.json', $authorListSchema);
$write('media.json', $mediaSchema);

// observations.md (lightweight)
$obs = "# Fake data observations (semantic-ex Phase 2)\n\n";
$obs .= "Generated from var/fake/*.json — 50 records per atomic entity.\n\n";
$logFields = [
    'Article' => [$articles, ['id','slug','title','body','excerpt','status','publishedAt','authorId','categoryId']],
    'Category' => [$categories, ['id','slug','name','description','parentId']],
    'Tag' => [$tags, ['id','slug','name']],
    'Author' => [$authors, ['id','name','email','bio']],
    'Media' => [$media, ['id','filename','mimeType','url','alt','width','height']],
];
foreach ($logFields as $label => [$rows, $names]) {
    $obs .= "## $label\n";
    foreach ($names as $f) {
        $vals = fields($rows, $f);
        $isInt = true;
        foreach ($vals as $v) {
            if ($v !== null && ! is_int($v)) {
                $isInt = false;
                break;
            }
        }

        if ($isInt) {
            $o = observeInt($vals);
            $obs .= sprintf("- `%s` (int): min=%d max=%d nulls=%d/%d\n", $f, $o['min'], $o['max'], $o['nulls'], count($vals));
        } else {
            $o = observeStr($vals);
            $obs .= sprintf("- `%s` (str): min_len=%d max_len=%d typical=%d nulls=%d/%d\n", $f, $o['min_len'], $o['max_len'], $o['typical_len'], $o['nulls'], count($vals));
        }
    }

    $obs .= "\n";
}

file_put_contents($fake . '/observations.md', $obs);

echo "Wrote schemas to {$out}\n";
foreach (scandir($out) as $f) {
    if ($f === '.' || $f === '..' || ! str_ends_with($f, '.json')) {
        continue;
    }

    echo "  $f\n";
}
